RecoPixelVertexing/PixelTrackFitting/src/PixelTrackCleanerBySharedHits.cc

```diff
@@ -25,49 +27,89 @@ void PixelTrackCleanerBySharedHits::cleanTracks(TracksWithTTRHs & trackHitPairs)
   unsigned int size = trackHitPairs.size();
   if (size <= 1) return;
 
-  auto kill = [&](unsigned int i) { delete trackHitPairs[i].first; trackHitPairs[i].first=nullptr;};
+  // sort (stabilize cleaning)
+  uint16_t score[size]; 
+  unsigned int ind[size];
+  for (auto i = 0U; i < size; ++i) {
+     ind[i]=i; 
+     score[i]= 32000-std::min(32000, int(trackHitPairs[i].first->chi2()*100.f));  // chi2: smaller is better
+     if (trackHitPairs[i].second.size()==4) score[i]+=32001; //  s4 always better than s3
+  }
+  std::sort(ind,ind+size,[&](unsigned int i, unsigned int j){return score[i]>score[j];});
+
+  auto kill = [&](unsigned int k) {delete trackHitPairs[k].first; trackHitPairs[k].first=nullptr;};
 
-  for (auto iTrack1 = 0U; iTrack1 < size; iTrack1++) {
+  // sorted: first is always better!
 
+  // first loop: only first    two hits....
```

Comment:
what is the advantage of doing 3 loops (first for 0,1 hits, then for 0,2 hits then for all) vs only one loop ?

Reply:
Because I tried and it happened to work! I have my own model why one has to make iterations fixing the layers (it is easy to prove that only one loop may not resolve ambiguities in "bundles" created by the large overlaps in the forward regions).
I did not try to optimize it: for instance the second loop may not add much.
Let's say: given the huge effect my simple change made, it is clear that 
1) this ambiguity solver need more work
2) we need a more radical approach at pattern recognition level (fishbone).
2a) I plan to work on this last point (first on GPU, then porting it to CPU together with the whole new doublets+CA implementation....)

Reply:
sorry for not being very clear :-(
what I meant is, aren't the two loops equivalent to a single loop with
```
      if ((recHits1[0] == recHits2[0]) and 
         ((recHits1[1] == recHits2[1]) or ((recHits1.size() > 2) and (recHits2.size() > 2) and (recHits1[2] == recHits2[2])))
      {
        kill(iTrack2);
      }
```
?

Reply:
strictly speaking not, because the first loop removes tracks that will not be tested in the second.
Because of the sorting this may now be less severe, still it is easy to make examples with just three tracks (say (1,2,3,7), (1,4,3,8), (1,4,5,9)) for which 2 loops will produce a different result than just one loop.

Reply:
To be clear, my solution is purely heuristic. no study done. It happens to work better than the previous implementation.

Reply:
> it is easy to make examples with just three tracks (say (1,2,3,7), (1,4,3,8), (1,4,5,9)) for which 2 loops will produce a different result than just one loop.

OK, I see what you mean: the logic is the same, but since the comparisons would be done in a different order the outcome could be different.